Add SVD reference design (#3)
KevinUTAT/oneAPI-samples_wenkaixu_fork: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/README.md

# `Singular Value Decomposition (SVD)` Sample
The `SVD` reference design demonstrates a singular value decomposition implementation for real matrices on an FPGA.

| Area                  | Description
|:---                   |:---
| What you will learn   | How to implement an FPGA version of singular value decomposition.
| Time to complete      | 1 hr (not including compile time)
| Category              | Reference Designs and End to End

## Purpose

> This FPGA reference design demonstrates the Singular Value Decomposition (SVD) of real matrices. SVD is a common linear algebra factorization used for applications such as matrix pseudo-inverse and solving homogeneous linear equations. The SVD of a real, 2D input matrix $A is defined as

Reviewer: ```suggestion
This FPGA reference design demonstrates the Singular Value Decomposition (SVD) of real matrices. SVD is a common linear algebra factorization used for applications such as matrix pseudo-inverse and solving homogeneous linear equations. The SVD of a real, 2D input matrix $A$ is defined as
```

> This FPGA reference design demonstrates the Singular Value Decomposition (SVD) of real matrices. SVD is a common linear algebra factorization used for applications such as matrix pseudo-inverse and solving homogeneous linear equations. The SVD of a real, 2D input matrix $A$ is defined as

Reviewer 2: ...defined as -> ...defined as follows:

```math
A = USV^T
```

where $U$ and $V$ are orthonormal bases of $A$, called (left and right) singular vectors. Orthonormal bases are vectors that are orthogonal to each other and of length 1.

$S$ is a diagonal matrix (matrix where all elements but the diagonals are zeros). The diagonals are called singular values corresponding to each singular vectors.

> SVD can also be applied to complex matrices, however this design currently doesn't support complex matrices for simplicity.

Reviewer 2: We can make this sentence a bit simpler:

While you can apply SVD to complex matrices, this design does not support complex matrices to keep the design simple.

> SVD operation can be applied to complex matrices, however this design currently don't support complex matrices for simplicity.

Reviewer: ```suggestion
SVD can also be applied to complex matrices, however this design currently doesn't support complex matrices for simplicity.
```

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
>
> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

### Performance

Performance results are based on testing as of March 20, 2024 with fixed 55 iterations.

Using BSP: Agilex_ref 23.3
> **Note**: Refer to the [Performance Disclaimers](/DirectProgramming/C++SYCL_FPGA/README.md#performance-disclaimers) section for important performance information.

| Device                                            | Throughput
|:---                                               |:---

> | Terasic DE10-Agilex Development Board             | 17.4k matrices/s for matrices of size 8 x 16

Reviewer: ```suggestion
| Terasic DE10-Agilex Development Board             | 17.4k matrices/s for matrices of size 8 x 16
```
do you mean size 16x8?

## Key Implementation Details
This SVD design consists of 4 computation kernels, as well as several memory access kernels to handle input and output. These kernels are connected through inter-kernel pipes and input/output through unified shared memory (USM).

![](assets/SVD.svg)

> 

Reviewer: There are a lot of source files in this design, please clarify which ones are for testbench/host code and which ones result in device code. ideally a table here to map the kernels shown in your diagram to the source files.

### Input covariance matrix computation
The covariance computation in this design is the same as used in the [PCA](../pca/README.md) reference design except without standardization.

The covariance of input A is equal to the transpose of A multiplied by A:
```math
Cov(A) = A^{T}A
```

> Therefore this kernel only performs a matrix multiplication by blocks.

Reviewer: ```suggestion
Therefore this kernel performs a matrix multiplication by blocks, as described in the PCA reference design.
```

### Eigenvalues and Eigenvectors computation
We are reusing the `fpga_linalg::StreamingEigen` design from the PCA sample to compute eigenvalues and eigenvectors of the input covariance matrix. These eigenvalues and eigenvectors are going to be used to construct the final outputs of SVD.

> This kernel only work with input of rank sufficient matrix (all columns are linearly independent). It produces an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag is set to 1, the result is incorrect.

Reviewer: ```suggestion
This kernel only works with input of rank sufficient matrix (all columns are linearly independent). It produces an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag is set to 1, the result of the SVD is known to be incorrect.
```

> This kernel only works with input of rank sufficient matrix (all columns are linearly independent). It produces an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag is set to 1, the result of the SVD is known to be incorrect.

Reviewer 2: ...only works... -> ...works only...

> The source code for the the Eigenvalues and Eigenvectors Computation kernel can be found in `streaming_eigen.hpp` in the shared [include](../../include) directory. An detailed explanation of the algorithm of computing Eigenvalues can be found [here](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.

Reviewer 2: Eigenvalues and Eigenvectors Computation

Why is this upper case?

> ### Construct output from Eigenvalues and Eigenvectors

Reviewer 2: output -> Output
(consisten title-case capitalization)

The final outputs of SVD can be constructed from the eigenvalues and eigenvectors of the input covariance. 

> Consider an input matrix A of size m x n.

Reviewer: ```suggestion
Consider an input matrix $A$ of size $m \times n$.
```

#### V Matrix

> The right singular vectors (the V matrix) can be produced by simply copying the eigenvectors , and should be of size n x n:

Reviewer: ```suggestion
The right singular vectors (the V matrix) can be produced by simply copying the eigenvectors , and should be of size $n \times n$:
```
i won't correct this anymore, you can do the rest 😊

> The right singular vectors (the $V$ matrix) can be produced by simply copying the eigenvectors , and should be of size $n \times n$:

Reviewer 2: remove "simply".

```math
S =
\begin{bmatrix}
\sqrt{\lambda_{0}} & 0                  & 0                  & ... & 0 \\
0                  & \sqrt{\lambda_{1}} & 0                  & ... & 0 \\
0                  & 0                  & \sqrt{\lambda_{2}} & ... & 0 \\
...                & ...                & ...                & ... & 0 \\
0                  & 0                  & 0                  & 0   & \sqrt{\lambda_{n}} \\
0                  & 0                  & 0                  & 0   & 0
\end{bmatrix} 
```

#### $U$ Matrix
The left singular vectors (the $U$ matrix) cannot be directly constructed with eigenvalues and eigenvectors of the input covariance. However, since we have now computed the $U$ and $S$ matrix, and knowing the original input $A$ matrix, we can calculate $U$ matrix using the original SVD relationship:
```math
A = USV^T
```
Since the $V$ matrix is orthogonal by construction, $V^{-1}$ is the same as $V^T$, therefore:
```math
AV = US
```
And since $S$ is a diagonal matrix, $S^{-1}$ is the same as multiplying by the reciprocals of each element on the main diagonal of $S$.

> So to sum up, the $m \times n$ portion of Matrix $U$ ($U_{[0:m][0:n]}$) can be obtained through:

Reviewer 2: So to sum up... -> In summary...

> So to sum up, the $m \times n$ portion of Matrix $U$ ($U_{[0:m][0:n]}$) can be obtained through:

Reviewer 2: Martix -> matrix

![](assets/U_matrix.png)

When the input matrix $A$ is not square, the number of eigenvectors calculated is less than the number of columns in the $U$ matrix. An orthogonalization kernel is needed to generate extra orthogonal vectors.

> ### $U$ Matrix orthogonalization

Reviewer 2: orthogonalization -> Orthogonalization

As mentioned above, when extra filler vectors are needed to complete the U matrix, we need to orthogonalize the matrix.

> An efficient algorithm to do this is already implemented in our [QR Decomposition sample](../qrd/README.md). So here we will just insert an instance of the `fpga_linalg::streamingQRD` design. The Q output of this kernel is orthogonalized U matrix. And the Q output is not used in the SVD design.

Reviewer: ```suggestion
An efficient algorithm to do this is already implemented in our [QR Decomposition sample](../qrd/README.md), so here we will insert an instance of the `fpga_linalg::streamingQRD` design. The $Q$ output of this kernel is orthogonalized $U$ matrix. Since we only care about the orthogonalized $U$ matrix, $R$ output of the Streaming QRD kernel is discarded.
```

> ### Demo Testbench

Reviewer: ```suggestion
### Demonstration Testbench
```

> ### Demo testbench

Reviewer 2: testbench -> Testbench

In this sample, a testbench is used to demonstrate the SVD design. The test bench uses an input matrix of size 16 x 8.

> The resulting singular values are compared against results obtained ahead of time using NumPy package in python.

Reviewer: ```suggestion
The resulting singular values in $S$ are compared against results obtained ahead of time using the NumPy package in Python <version>.
```
you should specify the python version and for thoroughness, also include the python code.

Since the singular vectors in $V$ are non-unique, their correctness are checked by (1) checking for orthogonality, and (2) that they satisfy the relationship $A = USV^T$.

> 

Reviewer: How about the $U$ matrix? or is this also checked by verifying $A=USV^T$?

> ## Build the `SVD` design

Reviewer 2: design -> Design

# `Singular Value Decomposition (SVD)` Sample
The `SVD` reference design demonstrates a singular value decomposition implementation for real matrices on an FPGA.

| Area                  | Description
|:---                   |:---
| What you will learn   | How to implement an FPGA version of singular value decomposition.

> | Time to complete      | 1 hr (not including compile time)

Reviewer: ```suggestion
| Time to complete      | 1 hr (not including compile time)
```

actually?

Author: I am not sure how to estimate this. The PCA and QRD both states 1 hour, I figure this probably takes similar amount of time.

Reviewer 3: This is super very bad metric as this varies widely depending on the user. We are forced but the code samples workgroup to include it.

| Category              | Reference Designs and End to End

## Purpose
This FPGA reference design demonstrates the Singular Value Decomposition (SVD) of real matrices. SVD is a common linear algebra factorization used for application such as matrix pseudo-inverse and solving homogeneous linear equations.

SVD usually described for factorization of a 2D input matrix A such that $A = USV^T$

> Where U and V are made of (left and right) singular vectors and are both orthogonal. And S is a diagonal matrix consisted of singular values corresponding to singular vectors in U and V.

Reviewer: ```suggestion
This FPGA reference design demonstrates the Singular Value Decomposition (SVD) of real matrices. SVD is a common linear algebra factorization used for applications such as matrix pseudo-inverse and solving homogeneous linear equations. The SVD of a real, 2D input matrix $A is defined as

    ```math
    A = USV^T
    ```

where $U$ and $V$ are made of (left and right) singular vectors and are both orthogonal. $S$ is a diagonal matrix consisting of singular values corresponding to singular vectors in $U$ and $V$.
```
(I indented the ```math section to make it work for the comment; feel free to un-indent it 😊)

when you refer to a named matrix such as *A* or *V*, I think it would be easier to read if you formatted it so it sticks out. If you can't use the LaTeX formatting, it's enough to *italicize*.

Reviewer: I'd like you to briefly define the linear algebra terms in the Implementation Details section.
* singular value
* singular vector
* orthogonal
* diagonal matrix

Also you should acknowledge that SVD can also apply to complex matrices, and explain why this design only shows SVD of real matrices.

| Optimized for        | Description
|:---                  |:---
| OS                   | Ubuntu* 20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10 <br> Windows Server* 2019
| Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
| Software             | Intel® oneAPI DPC++/C++ Compiler

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
>
> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

### Performance

> Performance results are based on testing as of March 20st, 2024.

Reviewer: ```suggestion
Performance results are based on testing as of March 20, 2024.
```

Reviewer: probably ought to mention what BSP was used as well, as this can impact performance numbers.

> **Note**: Refer to the [Performance Disclaimers](/DirectProgramming/C++SYCL_FPGA/README.md#performance-disclaimers) section for important performance information.

| Device                                            | Throughput
|:---                                               |:---
| Terasic DE10-Agilex Development Board             | 17.4k matrices/s for matrices of size 8 x 16

## Key Implementation Details

> This SVD design consist of 4 computation kernels (dark blue squares below), as well as several memory access kernels to handle input and output (light blue squares). These kernels are connected through inter-kernel pipes and input/output through unified shared memory (USM).

Reviewer: ```suggestion
This SVD design consists of 4 computation kernels (dark blue squares below), as well as several memory access kernels to handle input and output (light blue squares). These kernels are connected through inter-kernel pipes and input/output through unified shared memory (USM).
```

why not use accessors/buffers?

Author: Seems like USM is used across all of our linear alg samples?

![](assets/SVD.svg)

> 

Reviewer: When we were first talking about this design, i recall that you said we needed to re-run multiple iterations of the datapath before the SVD converged. I don't see any logic for that in this diagram, nor did I notice it in the code.

### Input covariance matrix computation
The covariance computation in this design is the same as used in the [PCA](../pca/README.md) reference design except without standardization.

> The covariance of input A is equal to transpose of A multiplies A:

Reviewer: ```suggestion
The covariance of input A is equal to the transpose of A multiplied by A:
```

```math
Cov(A) = A^{T}A
```
Therefore this kernel only performs a matrix multiplication by blocks.

> The implementation details of this computation is explained [here](../pca/README.md#input-matrix-standardization-and-covariance-matrix-computation) in the PCA sample.

Reviewer: ```suggestion
The implementation details of this computation is explained [here](../pca/README.md#input-matrix-standardization-and-covariance-matrix-computation) in the PCA sample.
```
If you are simply performing a matrix multiplication, why not use the `matmul` example? The description in the PCA code sample is entirely different.

Author: This is basically the "naive" matmul (with built in transposing) . Alternatively, using the systolic matmul with two copy of the input (one in row major, one in column major) should work, though i remember when I tested it, i couldn't get it work with the `streamingEign` kernel on my first try :(

Reviewer: for a large matrix we should be using the systolic matmul, as it gives us flexibility to scale resource usage somewhere between no unrolling, and totally unrolling the inner loop.

> ### Eigen values and Eigen vectors computation

Reviewer: ```suggestion
### Eigenvalues and Eigenvectors computation
```
eigenvalue and eigenvector are one word. Please fix other occurences.

We are reusing the `fpga_linalg::StreamingEigen` design from the PCA sample to compute eigen values and eigen vectors of the input covariance matrix. These eigen values and eigen vectors are going to be used to construct the final outputs of SVD.

> This kernel also produce an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag sets to 1, the result might be incorrect.

Reviewer: ```suggestion
This kernel also produces an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag is set to 1, the result might be incorrect.
```
what does 'might be incorrect' mean? How should the user check? (presumably by checking the relation A = USV*) Which result might be incorrect? U or V*?

Author: I am rewording this from 'might be incorrect' to 'is incorrect'

> The [sources code](../../include/streaming_eigen.hpp) for the design can be found in the include directory. An detailed explanation of the algorithm of computing eigen values can be found [here](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.

Reviewer: ```suggestion
The source code for the the Eigenvalues and Eigenvectors Computation kernel can be found in `streaming_eigen.hpp` in the shared [include](../../include) directory. An detailed explanation of the algorithm of computing eigen values can be found [here](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.
```

### Construct output from Eigen values and Eigen vectors
The final outputs of SVD can be constructed from the eigen values and eigen vectors of the input covariance. 

> Consider an SVD input A of size m x n.

Reviewer: ```suggestion
Consider an input matrix A of size m x n.
```

#### V Matrix

> The right singular vectors (the V matrix) can be produced by simply copy the eigen vectors , and should be of size n x n:

Reviewer: ```suggestion
The right singular vectors (the V matrix) can be produced by simply copying the eigenvectors, and should be of size n x n:
```

```math
V = [v_0, v_1, v_2, ...,v_n] 
```

#### S Matrix

> The singular value matrix (the S matrix) can be constructed as an m x n diagonal matrix (all but the diagonals are zeros) where the diagonals are square root of each eigen values:

Reviewer: ```suggestion
The singular value matrix (the S matrix) can be constructed as an m x n diagonal matrix (all but the values on the main diagonal are zeros) where the non-zero values are the square root of each eigenvalue:
```

```math
S =
\begin{bmatrix}
\sqrt{\lambda_{0}} & 0                  & 0                  & ... & 0 \\
0                  & \sqrt{\lambda_{1}} & 0                  & ... & 0 \\
0                  & 0                  & \sqrt{\lambda_{2}} & ... & 0 \\
...                & ...                & ...                & ... & 0 \\
0                  & 0                  & 0                  & 0   & \sqrt{\lambda_{n}} \\
0                  & 0                  & 0                  & 0   & 0
\end{bmatrix} 
```

#### U Matrix

> The left singular vector (the U matrix) cannot be directly constructed with eigen values and eigen vectors of the input covariance. However, since we have now computed the U and S matrix, and knowing the original input A matrix, we can calculate U matrix using the original SVD relationship:

Reviewer: ```suggestion
The left singular vector (the U matrix) cannot be directly constructed with eigen values and eigen vectors of the input covariance. However, since we have now computed the U and S matrix, and knowing the original input A matrix, we can calculate U matrix using the original SVD relationship:
```

> left singular vector (the U matrix)

Is it a vector or a matrix? 😊

```math
A = USV^T
```

> Since V matrix is orthogonal, inverse(V) is the same as transpose(V), therefore:

Reviewer: ```suggestion
Since the V matrix is orthogonal by construction, V^-1 is the same as V^T, therefore:
```

```math
AV = US
```

> And since S is a diagonal matrix, inverse(S) is the same as doing diagonal-element-wise reciprocals.

Reviewer: ```suggestion
And since S is a diagonal matrix, S^-1 is the same as multiplying by the reciprocals of each element on the main diagonal of S.
```

So to sum up, Matrix U can be obtained through:
![](assets/U_matrix.png)

> 

Reviewer: ```suggestion
when A is an nxn matrix... other constraints 😊
```

However, not all columns of the U matrix can be constructed this way. When the input matrix is not square, the number of eigen vectors calculated is less than the number of columns in U matrix.

> An orthogonalization kernel is needed to generate extra orthogonal vectors.

Reviewer: ```suggestion
When the input matrix A is not square, the number of eigenvectors calculated is less than the number of columns in the U matrix. An orthogonalization kernel is needed to generate extra orthogonal vectors.
```

### U Matrix orthogonalization
As mentioned above, when extra filler vectors are needed to complete the U matrix, we need to orthogonalize the matrix.

> An efficient algorithm to do this is already implemented in our [QR Decomposition sample](../qrd/README.md). So here we will just insert an instant of the `streamingQRD` design. The Q output of this kernel is orthogonalized U matrix.

Reviewer: ```suggestion
An efficient algorithm to do this is already implemented in our [QR Decomposition sample](../qrd/README.md). So here we will just insert an instance of the `StreamingQRD` design. The Q output of this kernel is orthogonalized U matrix.
```

what about the R matrix?

### Demo testbench

> In this sample, a testbench is used to demo the SVD design. The test bench use an input matrix of size 16 x 8, and its know singular values to compare for correctness.

Reviewer: ```suggestion
In this sample, a testbench is used to demonstrate the SVD design. The test bench uses an input matrix of size 16 x 8, and its know singular values to compare for correctness.
```
> its know singular values to compare for correctness.

please rephrase this, I don't know what it means.

> Since the singular vectors are non-unique, their correctness are checked by checking its orthogonality and the relationship $A = USV^T$

Reviewer: ```suggestion
Since the singular vectors in $V$ are non-unique, their correctness are checked by (1) checking for orthogonality, and (2) that they satisfy the relationship $A = USV^T$.
```

### On Linux*

1. Change to the sample directory.
2. Configure the build system for the Agilex® 7 device family, which is the default.

   ```
   mkdir build
   cd build
   cmake ..
   ```

   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```
   >
   > Alternatively, you can target an explicit FPGA board variant and BSP by using the following command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant> -DIS_BSP=1
   >  ```
  > **Note**: You can poll your system for available BSPs using the `aoc -list-boards` command. The board list that is printed out will be of the form
  > ```
  > $> aoc -list-boards
  > Board list:
  >   <board-variant>
  >      Board Package: <path/to/board/package>/board-support-package
  >   <board-variant2>
  >      Board Package: <path/to/board/package>/board-support-package

>   > ```

Reviewer: ```suggestion
  > **Note**: You can poll your system for available BSPs using the `aoc -list-boards` command. The board list that is printed out will be of the form
   > ```
   > $> aoc -list-boards
   > Board list:
   >   <board-variant>
   >      Board Package: <path/to/board/package>/board-support-package
   >   <board-variant2>
   >      Board Package: <path/to/board/package>/board-support-package
   > ```
```

should be aligned with the previous note

| Argument  | Description
|:---       |:---
| `<num>`   | (Optional) Specifies the number of times to repeat the decomposition. Its default value is **1** for the emulation and simulation flow and **16384** for the FPGA flow.

### On Linux

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./svd.fpga_emu
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./svd.fpga_sim
   ```

#### Run on FPGA

1. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   ./svd.fpga
   ```

### On Windows

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   svd.fpga_emu.exe
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   svd.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```

#### Run on FPGA

1. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   svd.fpga.exe

>    ```

Reviewer: I thought that the runtime on windows didn't work for full boards anymore?

Author: So we basically don't support Windows for full-acceleration any more ?

Reviewer: correct

| Category              | Reference Designs and End to End

## Purpose
This FPGA reference design demonstrates the Singular Value Decomposition (SVD) of real matrices. SVD is a common linear algebra factorization used for applications such as matrix pseudo-inverse and solving homogeneous linear equations. The SVD of a real, 2D input matrix $A is defined as

```math
A = USV^T
```

> where $U$ and $V$ are orthonormal bases of A, called (left and right) singular vectors. (Orthonormal bases are vectors that are orthogonal to each others and of length 1)

Reviewer: ```suggestion
where $U$ and $V$ are orthonormal bases of $A$, called (left and right) singular vectors. Orthonormal bases are vectors that are orthogonal to each other and of length 1.
```

please make sure named variables are formatted correctly
